docs/development/fleetlock/protocol.md

```diff
@@ -0,0 +1,95 @@
+---
+parent: Development
+nav_order: 3
+layout: default
+---
+
+# FleetLock protocol
+
+This document describes an HTTP-based protocol for orchestrating fleet-wide reboots, used by Zincati.
+It is modeled after a distributed counting semaphore with recursive locking and lock-ownership.
+
+## Overview
+
+This is a request-response protocol, where operations are always initiated by the client (i.e. Zincati).
+Each operation consists of a JSON payload sent as a POST request to the server.
+
+At an high level, the client can perform two operations:
+
+ * `RecursiveLock`: try to reserve (lock) a slot for rebooting
+ * `UnlockIfHeld`: try to release (unlock) a slot that it was previously holding
+
+Semaphore locks are owned, so that only the client that created a lock can release it. All operations are recursive, so that multiple unbalanced lock/unlock actions are allowed.
+
+## Client state-machine
+
+When a client starts it proceeds into either "initialization" or "finalization" state, based on system condition (i.e. whether a finalization is already in progress).
+In "**initialization**" state, the client tries to release any reboot slot it may have previously held.
+A successful unlock operation means that the client can proceed into its "**steady**" state and look for further updates.
+When an update is found and locally staged, the client proceed into its "**pre-reboot**" state and tries to lock a reboot slot.
+A successful lock operation means that the client can proceed into its "**finalization**" state and finalize a pending update, then reboot.
+
+## Requests
+
+### Endpoints
+
+All endpoints defined below are relative to a common deployment-specific base URL:
+
+ * `/v1/pre-reboot`: reserve/lock a reboot slot
+ * `/v1/steady-state`: release/unlock a reboot slot
+
+### Body
+
+All POST requests contain well-formed JSON body according to the following schema:
+
+ * `client_params` (object, mandatory)
```

Comment:
nit/idea: It would be nice to be able to link to a jsonschema for this as well. Verified this against the example listed later.

Example:
```json
{
  "$id": "https://github.com/coreos/zincati/....../client_params.schema.json",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "FleetLock Protocol",
  "definitions": {
    "client_params": {
      "type": "object",
      "required": [ "id", "group" ],
      "properties": {
        "id": {
            "type": "string",
            "minLength": 1,
            "description": "client identifier (e.g. node name or UUID)"
        },
            "group": {
            "type": "string",
            "minLength": 1,
            "pattern": "^[a-zA-Z0-9.-]+$",
            "description": "reboot-group of the client"
        }
      }
    }
  },
  "type": "object",
  "required": ["client_params"],
  "properties": {
    "client_params": {
        "$ref": "#/definitions/client_params"
    }
  }
}
```

Reply:
Indeed. I was planning to lift https://github.com/opencounter/terraform-fleet-lock-dynamodb/blob/master/json_schema.json at some point after merging the initial spec. Your schema looks nice and fully fledged too!
I was also planning to have https://github.com/poseidon/fleetlock/blob/master/examples/body.json too.

Both are separate file that can land once the directory is in place, so I'd not bundle them in this PR right now.